cassandra/src/main/java/org/hibernate/ogm/datastore/cassandra/CassandraDialect.java

```diff
@@ -0,0 +1,329 @@
+			BoundStatement boundStatement = new BoundStatement( preparedStatement );
+			for ( int i = 0; i < columnValues.length; i++ ) {
+				DataType dataType = preparedStatement.getVariables().getType( i );
+				boundStatement.setBytesUnsafe( i, dataType.serialize( columnValues[i], protocolVersion ) );
+			}
+			return session.execute( boundStatement );
+		}
+		catch (DriverException e) {
+			log.failToExecuteCQL( statement.getQueryString(), e );
+			return null;
+		}
+	}
+
+	@Override
+	public Tuple getTuple(EntityKey key, TupleContext tupleContext) {
+
+		Select select = select().all().from( quote( key.getTable() ) );
+		Select.Where selectWhere = select.where( eq( quote( key.getColumnNames()[0] ), QueryBuilder.bindMarker() ) );
+		for ( int i = 1; i < key.getColumnNames().length; i++ ) {
+			selectWhere = selectWhere.and( eq( quote( key.getColumnNames()[i] ), QueryBuilder.bindMarker() ) );
+		}
+
+		Object[] columnValues = key.getColumnValues();
+		ResultSet resultSet = bindAndExecute( columnValues, select );
+
+		if ( resultSet.isExhausted() ) {
+			return null;
+		}
+
+		Row row = resultSet.one();
+		Tuple tuple = new Tuple( new ResultSetTupleSnapshot( row, protocolVersion ) );
+		return tuple;
+	}
+
+	@Override
+	public Tuple createTuple(EntityKey key, TupleContext tupleContext) {
+		Map<String, Object> toSave = new HashMap<String, Object>();
+		toSave.put( key.getColumnNames()[0], key.getColumnValues()[0] );
+		return new Tuple( new MapTupleSnapshot( toSave ) );
+	}
+
+	@Override
+	public void insertOrUpdateTuple(EntityKey key, Tuple tuple, TupleContext tupleContext)
+			throws TupleAlreadyExistsException {
+
+		List<TupleOperation> updateOps = new ArrayList<TupleOperation>( tuple.getOperations().size() );
+		List<TupleOperation> deleteOps = new ArrayList<TupleOperation>( tuple.getOperations().size() );
+
+		for ( TupleOperation op : tuple.getOperations() ) {
+			switch ( op.getType() ) {
+				case PUT:
+					updateOps.add( op );
+					break;
+				case REMOVE:
+				case PUT_NULL:
+					deleteOps.add( op );
+					break;
+				default:
+					throw new HibernateException( "TupleOperation not supported: " + op.getType() );
+			}
+		}
+
+		if ( deleteOps.size() > 0 ) {
+
+			Delete.Selection deleteSelection = delete();
+			for ( TupleOperation tupleOperation : deleteOps ) {
+				deleteSelection.column( quote( tupleOperation.getColumn() ) );
+			}
+			Delete delete = deleteSelection.from( quote( key.getTable() ) );
+			Delete.Where deleteWhere = delete.where(
+					eq(
+							quote( key.getColumnNames()[0] ),
+							QueryBuilder.bindMarker()
+					)
+			);
+			for ( int i = 1; i < key.getColumnNames().length; i++ ) {
+				deleteWhere = deleteWhere.and( eq( quote( key.getColumnNames()[i] ), QueryBuilder.bindMarker() ) );
+			}
+
+			bindAndExecute( key.getColumnValues(), delete );
+		}
+
+		if ( updateOps.size() > 0 ) {
+
+			Insert insert = insertInto( quote( key.getTable() ) );
```

Comment:
Ah, so INSERT in Cassandra has UPSERT semantics, right? Probably worth adding a short note.